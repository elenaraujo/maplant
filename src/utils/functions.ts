import L from 'leaflet'

export const removeClassname = (classname: string) => {
  if (typeof window !== 'undefined') {
    const el = document.getElementsByClassName(classname)
    if (el.length > 0) el[0].classList.remove(classname)
  }
}

export const getSvgFromSlug = (slug: string) => {
  const icon = L.icon({
    iconUrl: `img/customMarkers/${slug}.svg`,
    iconSize: new L.Point(74, 74)
  })
  return icon
}
