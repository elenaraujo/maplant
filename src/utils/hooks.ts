import { useEffect, useState } from 'react'

export const useHeigth = () => {
  const [height, setHeight] = useState(400)

  useEffect(() => {
    const footer = document.getElementsByClassName('footer')
    const header = document.getElementsByClassName('header')

    const footerAndHeaderHeight =
      footer[0].clientHeight + header[0].clientHeight

    setHeight(window.innerHeight - footerAndHeaderHeight)
  }, [])

  return height
}

export const useWidth = () => {
  const [isMobile, setIsMobile] = useState(false)
  const [windowWidth, setWindowWidth] = useState(1020)

  useEffect(() => {
    setIsMobile(window.innerWidth < 940)
    setWindowWidth(window.innerWidth)
  }, [setIsMobile])

  return {
    isMobile,
    setIsMobile,
    windowWidth,
    setWindowWidth
  }
}
