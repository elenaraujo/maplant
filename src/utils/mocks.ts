export const markersMock = [
  {
    coordinates: {
      lat: -63.78248603116502,
      lng: -38.40820312500001
    },
    plant: {
      plantName: 'Arenaria',
      slug: 'arenaria',
      image: {
        url: 'https://media.graphassets.com/gUH5IBoPThS5UnOH93D9'
      }
    }
  },
  {
    coordinates: {
      lat: -68.78248603116502,
      lng: -44.40820312500001
    },
    plant: {
      plantName: 'PlantaLegal',
      slug: 'planta-legal',
      image: {
        url: 'https://media.graphassets.com/gUH5IBoPThS5UnOH93D9'
      }
    }
  }
]
