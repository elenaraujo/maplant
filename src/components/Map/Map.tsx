import { useState } from 'react'
import { MapContainer, TileLayer, Marker, Popup } from 'react-leaflet'

import { MapProps } from 'pages/white-orchard'

import AddMarker from 'components/AddMarker/AddMarker'
import Footer from 'components/Footer/Footer'
import Header from 'components/Header/Header'
import PopupContent from 'components/PopupContent/PopupContent'
import { MyLatLngLEAFLET } from 'components/MyLatLngLEAFLET/MyLatLngLEAFLET'

import { getSvgFromSlug } from 'utils/functions'

// Implementar verificação de pattern ao fazer cadastro de location
// para evitar valores diferentes de number positivo ou negativo
const Map = ({ markers, plants }: MapProps) => {
  const [clickedPlace, setClickedPlace] = useState({ lat: 0, lng: 0 })
  const [addModalOpen, setAddModalOpen] = useState(false)

  return (
    <>
      <Header hideWorlds />

      <AddMarker
        latLng={clickedPlace}
        addModalOpen={addModalOpen}
        setAddModalOpen={setAddModalOpen}
        plants={plants}
      />
      <MapContainer
        center={[-70, -65]}
        zoom={4}
        scrollWheelZoom={false}
        style={{ height: '100%', width: '100%', margin: 'auto' }}
        // [diagonal inferior direita]
        // [diagonal superior esquerda]
        maxBounds={[
          [-150, 45],
          [0, -180]
        ]}
      >
        <MyLatLngLEAFLET setClickedPlace={setClickedPlace} />

        <TileLayer
          tms
          // minZoom={4}
          maxZoom={6}
          attribution='&copy; <a href="https://github.com/witcher3map/witcher3map-maps">Witcher3map</a> contributors'
          url="https://raw.githubusercontent.com/witcher3map/witcher3map-maps/master/white_orchard/{z}/{x}/{y}.png"
        />

        {markers?.map(({ id, coordinates, plant }: MarkerProps) => {
          const { lat, lng } = coordinates
          const { plantName, slug, image } = plant

          return (
            <Marker
              key={`marker-${slug}-${lat}-${lng}`}
              position={[lat, lng]}
              title={plantName}
              icon={getSvgFromSlug(slug)}
            >
              <Popup maxWidth={250} minWidth={230} closeButton={false}>
                <PopupContent
                  markerId={id}
                  image={image}
                  plantName={plantName}
                  slug={slug}
                  plants={plants}
                />
              </Popup>
            </Marker>
          )
        })}

        {addModalOpen && (
          <Marker position={clickedPlace} title={'temporaryMarker'}>
            <Popup maxWidth={250} minWidth={230} closeButton={false}>
              Here is where your new marker will be registered
            </Popup>
          </Marker>
        )}
      </MapContainer>
      <Footer />
    </>
  )
}

export type MarkerProps = {
  id: string
  coordinates: { lat: number; lng: number }
  plant: {
    image: { url: string }
    plantName: string
    slug: string
  }
}

export default Map
