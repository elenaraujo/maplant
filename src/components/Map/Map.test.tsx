import { render, screen } from '@testing-library/react'
import Map from './Map'
import { markersMock } from 'utils/mocks'

describe('Map: ', () => {
  it('should render without any marker', () => {
    render(<Map markers={markersMock} />)

    expect(
      screen.getByRole('link', {
        name: /a javascript library for interactive maps/i
      })
    )
  })

  it('should render with the marker in correct place', () => {
    render(<Map markers={markersMock} />)
    expect(screen.getByTitle(/Arenaria/i)).toBeInTheDocument()
    expect(screen.getByTitle(/PlantaLegal/i)).toBeInTheDocument()
  })

  it('should render with the marker in correct place', () => {
    const { container } = render(<Map markers={markersMock} />)
    expect(container).toMatchSnapshot()
  })
})
