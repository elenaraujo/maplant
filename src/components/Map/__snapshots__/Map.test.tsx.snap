// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map:  should render with the marker in correct place 1`] = `
<div>
  <div
    class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    style="height: 100%; width: 106em; margin: auto; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 6; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="https://raw.githubusercontent.com/witcher3map/witcher3map-maps/master/white_orchard/4/5/3.png"
              style="width: 256px; height: 256px; left: -28px; top: -107px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      >
        <img
          alt=""
          class="leaflet-marker-shadow leaflet-zoom-hide"
          src="marker-shadow.png"
          style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 303px; top: -181px;"
        />
        <img
          alt=""
          class="leaflet-marker-shadow leaflet-zoom-hide"
          src="marker-shadow.png"
          style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 235px; top: -39px;"
        />
      </div>
      <div
        class="leaflet-pane leaflet-marker-pane"
      >
        <img
          alt="Marker"
          class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
          role="button"
          src="marker-icon.png"
          style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 303px; top: -181px; z-index: -181;"
          tabindex="0"
          title="Arenaria"
        />
        <img
          alt="Marker"
          class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
          role="button"
          src="marker-icon.png"
          style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 235px; top: -39px; z-index: -39;"
          tabindex="0"
          title="PlantaLegal"
        />
      </div>
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              −
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            Leaflet
          </a>
           
          <span
            aria-hidden="true"
          >
            |
          </span>
           © 
          <a
            href="https://github.com/witcher3map/witcher3map-maps"
          >
            Witcher3map
          </a>
           contributors
        </div>
      </div>
    </div>
  </div>
</div>
`;
